File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
const MOVES = { E: [0, 1], W: [0, -1], S: [1, 0], N: [-1, 0] };

function solution(park, routes) {
    let curR = 0;
    let curC = 0;
    const H = park.length;
    const W = park[0].length;

    for (let r = 0; r < H; r++) {
        if (park[r].includes('S')) {
            curR = r;
            curC = park[r].indexOf('S');
            break;
        }
    }
    
    for (const route of routes) {
        const [dir, distance] = route.split(" ");
        let [nr, nc] = [curR, curC];
        let isValid = true;

        for (let i = 0; i < Number(distance); i++) {
            nr += MOVES[dir][0];
            nc += MOVES[dir][1];

            if (nr < 0 || nr >= H || nc < 0 || nc >= W || park[nr][nc] === 'X') {
                isValid = false;
                break;
            }
        }
        
        if (isValid) {
            curR = nr;
            curC = nc;
        }
    }

    return [curR, curC];
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    r = 0
    c = 0
    
    H = len(park)
    W = len(park[0])

    for i in range(H):
        for j in range(W):
            if park[i][j] == 'S':
                r, c = i, j
                break
    
    directions = {"N": (-1, 0), "S": (1, 0), "E": (0, 1), "W": (0, -1)}
    
    for route in routes:
        direct, dist = route.split()
        dist = int(dist)
        dr, dc = directions[direct]
        
        is_valid = True
        nr, nc = r, c
        
        for _ in range(dist):
            nr += dr
            nc += dc
            if not (0 <= nr < H and 0 <= nc < W) or park[nr][nc] == 'X':
                is_valid = False
                break
        
        if is_valid:
            r, c = nr, nc
        
    return [r, c]